Unify and clean up POMs and activate integration test execution through profile

HannesWell opened the pull request:
This PR unifies and cleans up the pom.xml files of jzy3d as outlined in #153.

The clean-up mainly removes redundant elements and moves as much configration as possible towards the root POM. The root pom.xml contains a `dependencyManagement` section were the versions of all dependencies used anywhere in jzy3d are configured (but not referenced!). So if the version of any library needs to be updated it only needs to be done in the root pom.
The same is done for plug-ins used outside the root pom. But it could be changed that all plugins used are listed in the `pluginManagement` section.

Additionally the SWT dependencies are changed to the official ones from Eclipse at Maven-central.

The integration test are now executed when the profile `integration-tests` is activated. So to run the full build including integration tests executed the following command:
`mvn clean verify -Pintegration-tests`

The java-doc-, source- and test-jar generation is no done in the package phase (the default). If one does not want to execute it a phase before that can be seletec as target (like `mvn test`).

HannesWell commented:
When the Maven default life-cycle now runs up to the verify phase the java-doc generation is performed and print some warnings about not correctly escaped characters.
I can fix those warnings if you want them to vanish.

jzy3d commented:
Thank you this looks good ! I’ll try it ASAP on my computer.

For the warning as you like. I think they were already there before so we can address later.

HannesWell commented:
OK, great.

Yes they were there before but were hidden because the java-doc generation was attached to the deploy phase and therefore the java-doc was only generated during deployment.
But I think it is bad to do any part of the build only during deployment (except for the deployment itself of course) because you then deploy something you've never created/tested before. Or if you wanted to test it without deployment you had to disable the actual deployment (I assume that the commented-out configuration of the maven-deploy-plugin existed therefore). But this is a bit cumbersome.
And furthermore if source- and java-doc are build every time the build runs warnings are displayed immediately and can be fixed immediately and do not reveal themselves as a big group later.

HannesWell commented:
Once we have finalized this PR, I can add a follow up to update the build section in the `README.md` and the `workflows/maven.yml`. The latter should still work but contains unnecessary elements.

jzy3d commented:
Agree with the idea of verifying documentation at `install` phase despite it slow down usual build. May we attach to `verify` instead? If we can't nevermind, I agree that `deploy` must remain safe.

I suggest a last improvement on your PR : add yourself to the list of developers referenced in the main POM file :) thank you really a lot for your clear proposals and great contributions.

One point you may be interested in : I can successfully test HiDPI offscreen when project are setup for Java 9. However I really must stick on Java 8. For this I am thinking of adding maven module `${basedir}/tests/jzy3d-java9` containing a pom file only overriding the default Java 8 version of root pom file, and running all tests relevant to Java9+ that may be moved here. I think it is more readable and easy to do than creating maven profile for executing these tests inside the project the should really test (on may think of test naming convention such as ITTest_Java9_* for including/excluding such tests according to the compiler configuration). It would require more documentation and reading for the user. What do you think?

HannesWell commented:
> Agree with the idea of verifying documentation at `install` phase despite it slow down usual build. May we attach to `verify` instead? If we can't nevermind, I agree that `deploy` must remain safe.

Sure in general it could be bound to any phase we want, but I don't know if the compiled sources are required to attach sources or to generate java-doc. Anyway, the `verify` phase is definitively possible. Because it takes not a lot of additional runtime I think it is acceptable. As you have probably already noticed, the java-doc generation is now quiet and only prints errors and warnings, not a line for every handled file.

> I suggest a last improvement on your PR : add yourself to the list of developers referenced in the main POM file :) thank you really a lot for your clear proposals and great contributions.

Thank you very much, I am honored. It is a pleasure that my work is appreciated.

I've updated the commits accordingly, bound the java-doc and source code generation to the verify phase and added myself to the list of developers.

> One point you may be interested in : I can successfully test HiDPI offscreen when project are setup for Java 9. However I really must stick on Java 8. For this I am thinking of adding maven module `${basedir}/tests/jzy3d-java9` containing a pom file only overriding the default Java 8 version of root pom file, and running all tests relevant to Java9+ that may be moved here. I think it is more readable and easy to do than creating maven profile for executing these tests inside the project the should really test (on may think of test naming convention such as ITTest_Java9_* for including/excluding such tests according to the compiler configuration). It would require more documentation and reading for the user. What do you think?

The thing is that in Maven the configuration of a build plugin always applies to the entire module regardless if the plugin is configured by default or by a profile. A profile is more an override of the default configuration activated under certain circumstances (like being activated explicitly with the -P argument, the presence of a certain file or the use of a certain Java version). Therefore you can only compile the entire module with Java-8 or Java-9. So yes from your description I think a separate module dedicated to tests that require Java-9 is the better option.

jzy3d commented:
Hi,
Here is a preliminary example of prompting [Java 9 (HiDPI) test case](https://github.com/jzy3d/jzy3d-api/blob/feature/colorbarWidthAuto/jzy3d-tests-java9/tests.md).

jzy3d commented on pom.xml:
```diff
@@ -72,155 +68,284 @@
-		<version.mvn.javadoc>2.9.1</version.mvn.javadoc>
-		<version.mvn.release>2.5.3</version.mvn.release>
+		<version.java>1.8</version.java>
+		<version.swt>3.105.3</version.swt>
```
I am surprised by this : why did you keep SWT as a configurable dependency while all other have defaults in main pom file?

HannesWell replied:
This is because the handling of SWT dependencies is not straight forward because Eclipse-p2's dependency-handling and maven's dependency-handling is mixed. Therefore you have to specify a dependency to a specific platform (linux, windows, mac) to compile successfully. If you only specify the dependency org.eclipse.platform::org.eclipse.swt you only get an empty hull with nothing in it. So in order to compile the sources of jzy3d successfully but without forcing projects that depend on jzy3d's SWT native project to depend on all SWT platforms jzy3d does I specified them as optional dependency. 

I wrote some explanation about it into the `jzy3d-native-jogl-swt/pom.xml`.
And this requires multiple locations were the SWT version is required. Therefore I created a parameter, so they don't have to be updated all in case another SWT version is used one day.

jzy3d commented:
Hi, I am [almost done rebasing your work](https://github.com/jzy3d/jzy3d-api/tree/pomStructure) on latest changes. 

Running the new jzy3d-tests-java9 module is failing as follow. Do you have a suggestion?

```
[ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.05 s <<< FAILURE! - in org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout
[ERROR] org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal  Time elapsed: 0.033 s  <<< ERROR!
java.lang.NoClassDefFoundError: jgl/GL$PixelScaleListener
	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:37)
	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:28)
Caused by: java.lang.ClassNotFoundException: jgl.GL$PixelScaleListener
	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:37)
	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:28)
...
```

While `jgl.GL$PixelScaleListener` is defined. 

Eclipse is not complaining for build, but also fail with the same error while running the test. Switching back to the master branch makes this disappear, both in console AND Eclipse.

HannesWell commented:
> 
> Running the new jzy3d-tests-java9 module is failing as follow. Do you have a suggestion?
> 
> ```
> [ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.05 s <<< FAILURE! - in org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout
> [ERROR] org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal  Time elapsed: 0.033 s  <<< ERROR!
> java.lang.NoClassDefFoundError: jgl/GL$PixelScaleListener
> 	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:37)
> 	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:28)
> Caused by: java.lang.ClassNotFoundException: jgl.GL$PixelScaleListener
> 	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:37)
> 	at jzy3d.tests.java9@2.0.0-SNAPSHOT/org.jzy3d.tests.integration.ITTest_AxisLabelRotateLayout.whenAxisLabelOrientationNotHorizontal(ITTest_AxisLabelRotateLayout.java:28)
> ...
> ```
> 
> While `jgl.GL$PixelScaleListener` is defined.
> 
> Eclipse is not complaining for build, but also fail with the same error while running the test. Switching back to the master branch makes this disappear, both in console AND Eclipse.

I don't have an immediate answer, but I will look into it this evening. But Eclipse-JDT sometimes less strict and searches for dependencies in a wider area and therefore finds dependencies that are not found in the build.

HannesWell commented:
The problem should be fixed now. At least the test now fail with miss-matching sizes and not with the `NoClassDefFoundError` anymore (I could reproduce it while building it locally and for running the Unit-tests in Eclipse).
The problem was caused by a managed dependency to org.jzy3d::jGL in version 2.5. I'm sure anymore why I added it, but it was probably accidentally.

Removing the mentioned managed dependency fixed the problem as described initially.
I rebased the commits on top of the current master. While rebasing I resolved the merge-conflicts immediately, so there are no merge-conflict markers. Besides that I also removed unnecessary dependencies to mockito in child-modules and incorporated your clean-ups to the `jzy3d-tests-java9/pom.xml` in the corresponding commits.
Since I rebased the changes the `maven-surefire-plugin` element in the root pom's `build-plugins` section is again removed, assuming you did not re-added it intentionally and I think this is still not necessary. But please tell me if I'm wrong.
In case the inherited maven-surefire-plugin configuration causes trouble, it could be [skipped](https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#skip) for the java-9 tests.

The other differences between your and my branch are more or less cosmetic, but please compare to branches to see if these changes are fine.

jzy3d commented:
Thanks Hannes! I indeed move jGL from standalone project to child module in Jzy3D to make things easier. 

It's a pity you did the rebase : I did it as well and resolved all conflicts. I am not really used to github PR (I use gitlab at work and things are a little different). What I did on my side was importing your branch locally and rebasing it on master. I thought you would get these updates on your side, and that merging jzy3d/pomStructure would mark your branch HannesWell/jzy3d/pomStructure as merged. Am I wrong?

HannesWell commented:
Your welcome!

I've only seen the `jzy3d/pomStructure` branch in my repo (jzy3d is the upstream remote), but from all I can see this branch is not merged. And this PR is also still open for me.
For me doing the rebase again wasn't a problem and I took the change to resolve the conflicts immediately to avoid the need for a separate commit to resolve the conflicts. I tought a clean git-history would be nice here. And I also took the chance to incorporate clean-ups at locations that have changed in the meantime and to rebase on top of the latest changes.

So if you agree to the changes, you should be able to merge my new/forced pushed branch with the latest changes as a fast forward operation.
But if you prefer it, I can also create a commit on top of your `jzy3d/pomStructure` branch to fix the class-loading issue.

Unfortunately I don't know anything about the 'consumer' side of github-PR's I've only created them yet and at work I use Gerrit.

jzy3d commented:
Thanks. 

So now it is almost ok. There is a weird error when running ITTests with Failsafe on my Mac, so I just replaced by surefire.

https://github.com/jzy3d/jzy3d-api/blob/pomStructure/pom.xml#L493

I'll merge this and we'll be done.

HannesWell commented:
>
> So now it is almost ok. There is a weird error when running ITTests with Failsafe on my Mac, so I just replaced by surefire.
> 
That's indeed weired because from all I know the only difference between surefire and failsafe ist that the first fails the build immediately when the first test fails and the later runs all tests of the module and fails in the end if any test failed.

> https://github.com/jzy3d/jzy3d-api/blob/pomStructure/pom.xml#L493
This looks good to me and works at my computer.

> I'll merge this and we'll be done.
Great😄

jzy3d commented:
Congrats Hannes, this is merged! Thank you for your contribution and taking time to explain everything.

HannesWell commented:
Great. Thanks and you are welcome.
